### PR TITLE
Bug/event enhanced unique key

### PR DESCRIPTION
**Pull request by fivetran-catfritz**
## PR Overview
**This PR will address the following Issue/Feature:**
- #18 

**This PR will result in the following new package version:**

- v0.5.0 due to incremental and materialization changes.

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

> ## Breaking Changes
> Users should perform a `--full-refresh` when upgrading to ensure all changes are applied correctly. This includes updates to unique key generation, materialization, and incremental strategies, which may affect existing records.
> 
> - Revised `unique_key` generation for `amplitude__event_enhanced` using `unique_event_id` and `unique_event_type_id` to prevent duplicate records.
>     - The unique key was previously generated from `unique_event_id` and `event_day`, which caused duplicate keys for some users and prevented incremental runs.
> - Made the `int_amplitude__date_spine` materialization ephemeral to reduce the number of tables and simplify incremental model dependencies.
> - Updated incremental loading strategies:
>   - **BigQuery** and **Databricks**: `insert_overwrite` for compute efficiency
>   - **Snowflake**, **Redshift**, and **Postgres**: `delete+insert`
> 
> ## Features
> - Added a default 3-day lookback period for incremental models to handle late-arriving records. Customize the lookback duration by setting the `lookback_window` variable in `dbt_project.yml`. For more information, refer to the [Lookback Window section of the README](https://github.com/fivetran/dbt_amplitude/blob/main/README.md#lookback-window).
> - Added the `amplitude_lookback` macro to simplify lookback calculations across models.
> 
> ## Documentation updates
> - Updated outdated or missing field definitions in dbt documentation.
> 
> ## Under the hood
> - Adjusted the `event_time` field in the `event_data` seed file to ensure records are not automatically excluded during test runs.
> - Added consistency tests for end models.

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [x] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
- [x] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:

1. See internal ticket for  detailed test data validation
2. Consistency tests pass:
Screenshot 2024-11-04 at 7 37 38 PM

### If you had to summarize this PR in an emoji, which would it be?
 
:dancer:

### CODE REVIEW

File: models/amplitude__daily_performance.sql [hunk 1]
```diff
@@ -12,29 +13,20 @@ with event_enhanced as (
...
-        , max_date
-        where event_day >= max_date.max_event_day
-
-    {% endif %}
+    select
+        distinct event_enhanced.event_type,
+        spine.date_day as event_day
+    from {{ ref('int_amplitude__date_spine') }} as spine
+    join event_enhanced -- this join limits the incremental run
+        on spine.date_day >= event_enhanced.event_day -- each event_type will have a record for every day since their first day
```
Comment (fivetran-catfritz) [inline models/amplitude__daily_performance.sql:29]: This was moved here from the original `int_amplitude__date_spine` so the date_spine could be ephemeral and also to reduce the number of pings to `amplitude__event_enhanced`.

File: models/intermediate/int_amplitude__date_spine.sql [hunk 2]
```diff
@@ -1,77 +1,24 @@
...
 
+    select
+        cast(spine.date_day as date) as date_day
     from (
         {{ dbt_utils.date_spine(
             datepart = "day", 
-            start_date =  "cast('" ~ var('amplitude__date_range_start',  '2020-01-01') ~ "' as date)", 
-            end_date = "cast('" ~ var('amplitude__date_range_end',  end_date_adjust) ~ "' as date)" 
+            start_date =  "cast('" ~ var('amplitude__date_range_start',  '2020-01-01') ~ "' as date)",
+            end_date = "cast('" ~ var('amplitude__date_range_end', end_date_adjust) ~ "' as date)"
```
Comment (fivetran-catfritz) [inline models/intermediate/int_amplitude__date_spine.sql:17]: I thought about modernizing the date ranges, however they are directly tied to [these staging model filters](https://github.com/fivetran/dbt_amplitude_source/blob/main/models/stg_amplitude__event.sql#L76-L77), so it is out of scope for this update. I'm not even sure if it must be updated. If so, I can open an FR.

**Review by fivetran-catfritz — COMMENTED** [review]
@fivetran-joemarkiewicz Thanks! Made the changes, so this is ready for re-review.

**Comment by fivetran-reneeli** [comment]
Hi @fivetran-catfritz , the changes mainly look good but had some questions:
1. default lookback window: I was curious how the default was decided to be 3 days. I noticed in other packages it's usually 7. For example: [Shopify](https://github.com/fivetran/dbt_shopify/blob/903e5629b4afd3b47af5caffcb69196e78e7106f/models/shopify__discounts.sql#L20)

2. I'm still a little confused how the initial `unique_event_id` + `event_day` wasn't a unique_key for `amplitude__event_enhanced`. Especially with Amplitude's [deduping](https://amplitude.com/docs/get-started/track-your-progress):

> Amplitude de-duplicates your data so it doesn't log unique events multiple times. Amplitude checks the event ID, client event time, and device ID for every event. If the event isn't in the database, Amplitude writes it; otherwise, Amplitude drops the event.
> 
> If you're using the Amplitude HTTP API, add an [insert_id field](https://amplitude.com/docs/apis/analytics/http-v2). Amplitude ignores subsequent events sent with the same event ID/client event time/device ID or insert_id within the past seven days.

But your comment does make sense, about how `event_type` is being left joined to `event`, so hence why `unique_event_type_id` should be brought in. When I initially made `amplitude__event_enhanced`, it was supposed to be at the unique_event_id grain. But looking back at the ERD, `event` to `event_type` is one to many, so I see how the fanout could occur and therefore the unique_key needs to account for that. Lmk if that's the conclusion you got.

**Comment by fivetran-catfritz** [comment]
@fivetran-reneeli My answers!
1. It ended up 3 because I was looking at fivetran_log's audit table when setting this up. 3 would mean less data to process, but will take into account fewer late arrivals. If you think it should be 7 I can change it. 
2. Yes exactly it's because of the one to many and the left join.

**Review by fivetran-reneeli — APPROVED** [review]
looks good!

**Comment by fivetran-reneeli** [comment]
Thanks @fivetran-catfritz ! Approved

**Pull request merged**